src/api/procedures/removeExternalAgent.ts

```diff
@@ -0,0 +1,119 @@
+import { TxTags } from 'polymesh-types/types';
+
+import { CustomPermissionGroup } from '~/api/entities/CustomPermissionGroup';
+import { KnownPermissionGroup } from '~/api/entities/KnownPermissionGroup';
+import { Agent, PolymeshError, Procedure, SecurityToken } from '~/internal';
+import { ErrorCode, PermissionGroupType } from '~/types';
+import { ProcedureAuthorization } from '~/types/internal';
+import { stringToIdentityId, stringToTicker } from '~/utils/conversion';
+import { getDid } from '~/utils/internal';
+
+export interface RemoveExternalAgentParams {
+  target: string | Agent;
+}
+
+/**
+ * @hidden
+ */
+export type Params = RemoveExternalAgentParams & {
+  ticker: string;
+};
+
+/**
+ * @hidden
+ */
+export interface Storage {
+  token: SecurityToken;
+}
+
+/**
+ * @hidden
+ */
+const isFullGroupType = (group: KnownPermissionGroup | CustomPermissionGroup): boolean =>
+  group instanceof KnownPermissionGroup && group.type === PermissionGroupType.Full;
+
+/**
+ * @hidden
+ */
+export async function prepareRemoveExternalAgent(
+  this: Procedure<Params, void, Storage>,
+  args: Params
+): Promise<void> {
+  const {
+    context: {
+      polymeshApi: {
+        tx: { externalAgents },
+      },
+    },
+    context,
+    storage: { token },
+  } = this;
+
+  const { ticker, target } = args;
+
+  const [currentAgents, did] = await Promise.all([
+    token.permissions.getAgents(),
+    getDid(target, context),
+  ]);
+
+  const agentWithGroup = currentAgents.find(({ agent: { did: agentDid } }) => agentDid === did);
+
+  if (!agentWithGroup) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'The target Identity is not an External Agent',
+    });
+  }
+
+  if (isFullGroupType(agentWithGroup.group)) {
+    const fullGroupAgents = currentAgents.filter(({ group }) => isFullGroupType(group));
+    if (fullGroupAgents.length === 1) {
+      throw new PolymeshError({
+        code: ErrorCode.ValidationError,
+        message:
+          'The agent to remove is the last one assigned to the group of all transactions authorized',
```

Comment:
The target is the last agent with full permissions for this Security Token. There should always be at least one agent with full permissions